### PR TITLE
Added DPM-Solver++ sampling thanks to Katherine Crowson

### PR DESCRIPTION
**Pull request by Mut1nyJD**

### CODE REVIEW

File: denoising_diffusion_pytorch/elucidated_diffusion.py [hunk 1]
```diff
@@ -199,6 +199,31 @@ def sample(self, batch_size = 16, num_sample_steps = None, clamp = True):
         images = images.clamp(-1., 1.)
         return unnormalize_to_zero_to_one(images)
 
+    # thanks to Katherine Crowson (https://github.com/crowsonkb) for figuring it all out!
+    @torch.no_grad()
+    def sanple_using_dpmpp(self,batch_size = 16, num_sample_steps = None):
```
Comment (lucidrains) [inline denoising_diffusion_pytorch/elucidated_diffusion.py:204]: typo: `sanple`

File: denoising_diffusion_pytorch/elucidated_diffusion.py [hunk 2]
```diff
@@ -199,6 +199,31 @@ def sample(self, batch_size = 16, num_sample_steps = None, clamp = True):
...
+       for i in range(len(sigmas) - 1):
+           denoised = self.preconditioned_network_forward(images, sigmas[i].item())
+           t, t_next = t_fn(sigmas[i]), t_fn(sigmas[i + 1])
+           h = t_next - t
+           if old_denoised is None or sigmas[i + 1] == 0:
+               images = (sigma_fn(t_next) / sigma_fn(t)) * images - (-h).expm1() * denoised
+           else:
+               h_last = t - t_fn(sigmas[i - 1])
+               r = h_last / h
+               denoised_d = (1 + 1 / (2 * r)) * denoised - (1 / (2 * r)) * old_denoised
```
Comment (lucidrains) [inline denoising_diffusion_pytorch/elucidated_diffusion.py:221]: ohh so some sort of exponential moving average?

**Comment by lucidrains** [comment]
@Mut1nyJD thank you for this!

**Comment by lhaippp** [comment]
wow, that's awesome! Thanks again for your contribution! @lucidrains @Mut1nyJD

**Comment by Mut1nyJD** [comment]
> oh, could you add to the citations part of the readme too?

Sure thing!

**Comment by lucidrains** [comment]
@Mut1nyJD @BAUTAJD thank you!

**Pull request merged**